### PR TITLE
Store refund counter in EVM

### LINKED ISSUE #611: SstoreCallToSelfSubRefundBelowZero failing in Istanbul
The [SstoreCallToSelfSubRefundBelowZero](https://github.com/ethereum/tests/blob/52d5247fe41f489c274e4254b151520bc3f20cc7/src/GeneralStateTestsFiller/stSStoreTest/SstoreCallToSelfSubRefundBelowZeroFiller.json) test case is failing for Istanbul (see #607). Here's why:

The test case calls a contract which has an existing storage slot. It clears the slot, thereby refunding some gas according to EIP-2200. It then calls itself and in the call it sets the slot again to some value, which means the prior refund has to be cleared:

https://github.com/ethereumjs/ethereumjs-vm/blob/00d8fe3c7941fe571581fff1cd4e327980928127/lib/evm/opFns.ts#L1018-L1023

But here it panics because the refund counter of the `EEI` is 0 in the current message, i.e. in the nested call EEI is not aware that there has been some previous refund, because we create a new EEI instance for every message. We should move the refund counter to outer layers (two of the possibilities are in `EVM` or `StateManager`).

Similar to geth which store their refund counter in [StateDB](https://github.com/ethereum/go-ethereum/blob/a308f012ba6bac3249ae52727e1c52273390d7e9/core/state/statedb.go#L82) I tried to store it in our StateManager and could make the test pass. I also had to clear the counter in `runTx` for next txes. This raises the question whether we should also clear the counter in `runCall` and `runCode`? not sure...

I'll also try with storing it in `EVM`, because it somehow feels more logically related.

### PR DESCRIPTION
**Pull request by s1na**
To be merged into #607

Fixes #611

### CODE REVIEW

**Comment by s1na** [comment]
Hm, should we create separate PRs for istanbul bugfixes or do everything in #607? the thing is that if we create separate PRs, the CI will be failing for them (as other bugs are still there).

**Comment by holgerd77** [comment]
@s1na Think it makes sense to keep it towards #607 since there would be various things to address before CI will pass through again. Not sure if its realistic, but can we optimally very much prioritize here (Sina, could you eventually take some short-term extra time)? The HF on mainnet is in 3-4 weeks and if somehow possible we should get out a release out until mid next week?

@evertonfraga Can you also eventually also jump in here a bit? If you are not deep enough in the VM code yet some preparatory documentation work would also help very much, e.g. an analysis together with some listing which tests are actually failing and which related EIPs are affected, eventually you can also already dig in one level deeper and investigate what constellation these tests are actually testing.

//cc @alcuadrado 1-2 extra days as well? 🙂 Still have to ask you on your current team status, will continue conversation on Gitter.

File: lib/evm/eei.ts [hunk 1]
```diff
@@ -33,7 +33,6 @@ export interface Env {
 export interface RunResult {
   logs: any // TODO: define type for Log (each log: [Buffer(address), [Buffer(topic0), ...]])
   returnValue?: Buffer
-  gasRefund: BN
```
Comment (alcuadrado) [inline lib/evm/eei.ts]: Wouldn't this introduce a breaking change? I think we should duplicate the `gasRefund` value for now, and add a comment about this field being deprecated and removed in the future.

Is this possible @s1na? I'm not completely sure atm.

**Comment by evertonfraga** [comment]
@holgerd77 sure thing

**Pull request merged**